File: interview/Cracking the Coding Interview/chapter 2/2.7/c++/2.7.cpp
#include <iostream>

using namespace std;

template <typename T>
class Node {
	public:
	Node () { this->next = 0; }
	Node (T data, Node<T>* next) {
		this->data = data;
		this->next = next;
	}
	
	T data;
	Node<T> * next;
};

template <typename T>
class LinkedList {
	public:
	LinkedList() { this->head = 0;}
	LinkedList(Node<T>* head) { this->head  = head;}
	
	void insert(T data) {
		Node<T>* n = new Node<T>(data, head);
		head = n;
	}
	
	Node<T>* search(T data) {
		Node<T> * curr = head;
		
		while (curr && curr->data != data) {
			curr = curr->next;
		}
		
		return curr;
	}
	
	bool isPalindrome() {
		LinkedList<T> reverseList = LinkedList<T>();
		Node<T>* curr = head;
		while (curr) {
			reverseList.insert(curr->data);
			curr = curr->next;
		}
		
		Node<T>* curr1 = head;
		Node<T>* curr2 = reverseList.head;
		while (curr1 && curr2)
		{
			if (curr1->data != curr2->data)
				return false;
			
			curr1 = curr1->next;
			curr2 = curr2->next;
		}
		
		return true;
	}
	
	void print() {
		Node<T>* curr = head;
		while (curr){
			cout << curr->data << " ";
			curr = curr->next;
		}
		cout << endl;
	}
	Node<T>* head;
	
	
};


int main() {
	LinkedList<char> L = LinkedList<char>();
	
	L.insert('r');
	L.insert('a');
	L.insert('c');
	L.insert('e');
	L.insert('c');
	L.insert('a');
	L.insert('r');
	L.print();
	cout << L.isPalindrome() << endl;
	
	
	L = LinkedList<char>();
	
	L.insert('r');
	L.insert('a');
	L.insert('c');
	L.insert('e');
	L.insert('c');
	L.insert('a');
	L.print();
	cout << L.isPalindrome();
	return 0;
}

File: interview/Cracking the Coding Interview/chapter 2/2.7/python/2.7.py
class Node(object):
    def __init__(self, data=None, next_node=None):
        self.data = data
        self.next_node = next_node

    def get_data(self):
        return self.data

    def get_next(self):
        return self.next_node

    def set_next(self, new_next):
        self.next_node = new_next

class LinkedList(object):
    def __init__(self, head=None):
        self.head = head

    def insert(self, data):
        new_node = Node(data)
        new_node.set_next(self.head)
        self.head = new_node

    def size(self):
        current = self.head
        count = 0
        while current:
            count += 1
            current = current.get_next()
        return count

    def __len__(self):
        return self.size()

    def search(self, data):
        current = self.head
        found = False
        while current and found is False:
            if current.get_data() == data:
                found = True
            else:
                current = current.get_next()
        if current is None:
            raise ValueError("Data not in list")
        return current

    def __contains__(self, data):
        return self.search(data)
        
    def delete(self, data):
        current = self.head
        previous = None
        found = False
        while current and found is False:
            if current.get_data() == data:
                found = True
            else:
                previous = current
                current = current.get_next()

        if current is None:
            raise ValueError("Data not in list")
        if previuos is None:
            self.head = current.get_next()
        else:
            previous.set_next(current.get_next())

    def __repr__(self):
        s = ''
        current = self.head
        while current:
            s += str(current.get_data()) + ','
            current = current.get_next()
        return s[:-1]

    def isPalindrome(self):
        L = LinkedList()
        c = self.head
        while c != None:
            L.insert(c.get_data())
            c = c.get_next()

        c1 = self.head
        c2 = L.head
        while c1 != None and c2 != None:
            if c1.get_data() != c2.get_data():
                return False
            
            c1 = c1.get_next()
            c2 = c2.get_next()

        return True

L = LinkedList(Node('r',Node('a',Node('c',Node('e',Node('c',Node('a',Node('r'))))))))
print(L.isPalindrome())

L = LinkedList(Node('a',Node('b',Node('c',Node('d')))))
print(L.isPalindrome())
